layers/layers.go

```diff
@@ -0,0 +1,273 @@
+}
+
+// Layer represents a read only layer
+type Layer interface {
+	TarStreamer
+	Address() LayerAddress
+	Parent() (Layer, error)
+	Size() (int64, error)
+}
+
+// RWLayer represents a layer which is
+// read and writable
+type RWLayer interface {
+	TarStreamer
+	Path() (string, error)
+	Parent() (Layer, error)
+}
+
+type LayerStore interface {
+	Register(io.Reader, LayerAddress) (Layer, error)
+	Get(LayerAddress) (Layer, error)
+	Put(LayerAddress) error
+
+	Mount(id string, parent LayerAddress) (RWLayer, error)
+	Unmount(id string) error
+}
+
+type tarStreamer func() (io.Reader, error)
+
+type cacheLayer struct {
+	tarStreamer
+	address LayerAddress
+	parent  *cacheLayer
+	cacheID string
+	size    int64
+}
+
+func (cl *cacheLayer) TarStream() (io.Reader, error) {
+	return cl.tarStreamer()
+}
+
+func (cl *cacheLayer) Address() LayerAddress {
+	return cl.address
+}
+
+func (cl *cacheLayer) Parent() (Layer, error) {
+	return cl.parent, nil
+}
+
+func (cl *cacheLayer) Size() (int64, error) {
+	return cl.size, nil
+}
+
+type mountedLayer struct {
+	tarStreamer
+	mountID string
+	parent  *cacheLayer
+	path    string
+}
+
+func (ml *mountedLayer) TarStream() (io.Reader, error) {
+	return ml.tarStreamer()
+}
+
+func (ml *mountedLayer) Path() (string, error) {
+	return ml.path, nil
+}
+
+func (ml *mountedLayer) Parent() (Layer, error) {
+	return ml.parent, nil
+}
+
+type layerStore struct {
+	root   string
+	driver graphdriver.Driver
+
+	layerMap map[LayerAddress]*cacheLayer
+	layerL   sync.Mutex
+
+	mounts map[string]*mountedLayer
+	mountL sync.Mutex
+}
+
+func NewLayerStore(root string, driver graphdriver.Driver) (LayerStore, error) {
+	ls := &layerStore{
+		root:     root,
+		driver:   driver,
+		layerMap: map[LayerAddress]*cacheLayer{},
+		mounts:   map[string]*mountedLayer{},
+	}
+
+	// TODO: Load existing layers and references
+
+	return ls, nil
+}
+
+func (ls *layerStore) Register(ts io.Reader, parent LayerAddress) (Layer, error) {
+	var pid string
+	var p *cacheLayer
+	if string(parent) != "" {
+		l, ok := ls.layerMap[parent]
+		if !ok {
+			return nil, ErrLayerDoesNotExist
+		}
+		p = l
+		pid = l.cacheID
+	}
+
+	// Create new cacheLayer
+	layer := &cacheLayer{
+		parent:  p,
+		cacheID: stringid.GenerateRandomID(),
+	}
+
+	if err := ls.driver.Create(layer.cacheID, pid); err != nil {
+		return nil, err
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			logrus.Debugf("Cleaning up layer %s: %v", layer.cacheID, err)
+			if err := ls.driver.Remove(layer.cacheID); err != nil {
+				logrus.Errorf("Error cleaning up cache layer %s: %v", layer.cacheID, err)
+			}
+		}
+	}()
+
+	digester := digest.Canonical.New()
+	tr := io.TeeReader(ts, digester.Hash())
+
+	layer.size, err = ls.driver.ApplyDiff(layer.cacheID, pid, archive.Reader(tr))
+	if err != nil {
+		return nil, err
+	}
+
+	layer.tarStreamer = func() (io.Reader, error) {
+		archiver, err := ls.driver.Diff(layer.cacheID, pid)
+		return io.Reader(archiver), err
+	}
```

Comment:
I think after our discussion it makes sense to move these back to to layer store object